### ISSUE TITLE
TypeLoadException on an implementation of an abstract class

### ISSUE THREAD

**Issue by HB-2012** [RSuter/NSwag#1175]
I'm using NSwag to automatically generate a swagger.json and a C# Client at compile time. I'm running an ASP.NET Core 2 application, but I'm getting the following exception when I run the nswag command:
```
System.TypeLoadException: Method 'StartAsync' in type 'SuperCoolNamespace.JobService' from assembly 'SuperCoolAssembly, Version=0.1.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.

Server stack trace:
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at NSwag.Commands.SwaggerGeneration.WebApiToSwaggerCommand.<GetControllerTypesAsync>d__91.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\SwaggerGeneration\WebApiToSwaggerCommand.cs:line 285
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.SwaggerGeneration.WebApiToSwaggerCommand.<RunIsolatedAsync>d__88.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\SwaggerGeneration\WebApiToSwaggerCommand.cs:line 188
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.IsolatedCommandBase`1.IsolatedCommandAssemblyLoader`1.Run(String commandType, String commandData, String[] assemblyPaths, String[] referencePaths) in C:\projects\nswag\src\NSwag.Commands\Commands\IsolatedCommandBase.cs:line 59
   at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NSwag.Commands.IsolatedCommandBase`1.IsolatedCommandAssemblyLoader`1.Run(String commandType, String commandData, String[] assemblyPaths, String[] referencePaths)
   at NSwag.Commands.IsolatedCommandBase`1.<>c__DisplayClass13_0.<RunIsolatedAsync>b__0() in C:\projects\nswag\src\NSwag.Commands\Commands\IsolatedCommandBase.cs:line 49
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.IsolatedCommandBase`1.<RunIsolatedAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.IsolatedSwaggerOutputCommandBase.<RunAsync>d__4.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\IsolatedSwaggerOutputCommandBase.cs:line 24
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.NSwagDocumentBase.<GenerateSwaggerDocumentAsync>d__40.MoveNext() in C:\projects\nswag\src\NSwag.Commands\NSwagDocumentBase.cs:line 266
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.NSwagDocument.<ExecuteAsync>d__8.MoveNext() in C:\projects\nswag\src\NSwag.Commands\NSwagDocument.cs:line 76
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.Document.ExecuteDocumentCommand.<ExecuteDocumentAsync>d__9.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\Document\ExecuteDocumentCommand.cs:line 77
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.Document.ExecuteDocumentCommand.<RunAsync>d__8.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\Document\ExecuteDocumentCommand.cs:line 43
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.<ProcessSingleAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.<ProcessAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.Process(String[] args, Object input)
   at NSwag.Commands.NSwagCommandProcessor.Process(String[] args) in C:\projects\nswag\src\NSwag.Commands\NSwagCommandProcessor.cs:line 54
```

Here's some of the code I'm using. StartAsync is implemented in [HostedService](https://gist.github.com/davidfowl/a7dd5064d9dcf35b6eae1a7953d615e3), thus the error message is confusing me.

JobService:
```
public class JobService : HostedService
    {
        private ILogger _logger;
        public JobService(ILogger<JobService> logger)
        {
            _logger = logger;
        }

        protected override async Task ExecuteAsync(CancellationToken cancellationToken)
        {
            while(!cancellationToken.IsCancellationRequested)
            {
                Stopwatch stopwatch = new Stopwatch();
                _logger.LogDebug("Job Service started doing things...");

                _logger.LogDebug("Job Service completed {0}ms ({1} sec).", 
                    stopwatch.ElapsedMilliseconds, stopwatch.Elapsed.TotalSeconds);

                await Task.Delay(TimeSpan.FromMinutes(1)); // Run only once every minute.
            }
        }
    }
```

I feel like I'm doing it right. Any feedback would be appreciated.

**Comment by RSuter** [comment]
Can you provide a sample project?

**Comment by HB-2012** [comment]
Thanks for the quick reply. I'll do that.

**Comment by RSuter** [comment]
Just tried it with version v11.13.3, with the .NET Core 2.0 runtime: 

1. Update NSwag to v11.13.3 (e.g. `npm i nswag -g`)
2. Run with .NET Core 2.0 runtime: `nswag run /runtime:NetCore20`

**Comment by RSuter** [comment]
![image](https://user-images.githubusercontent.com/2603405/35852463-3fd9d096-0b2b-11e8-8691-b5844b9ac706.png)

**Comment by HB-2012** [comment]
Still no luck for me after updating NSwag. What version of dotnet are you running? I'm on 2.1.4, which seems to be the [latest version](https://github.com/dotnet/cli/releases).

**Comment by RSuter** [comment]
It's important to run nswag in the .NET Core runtime with `/runtime:NetCore20`, did you do that?

**Comment by HB-2012** [comment]
Yes I did. It does not appear to make a difference on my end. I get the exact same stack trace (according to KDiff3) with and without specifying the runtime.

```
PS C:\Users\joeri\source\repos\nswag1175\nswag1175> dotnet --version
2.1.4
PS C:\Users\joeri\source\repos\nswag1175\nswag1175> nswag version
NSwag command line tool for .NET 4.6.1+ WinX64, toolchain v11.13.3.0 (NJsonSchema v9.10.22.0 (Newtonsoft.Json v9.0.0.0))
Visit http://NSwag.org for more information.
NSwag bin directory: C:\Program Files (x86)\Rico Suter\NSwagStudio\Win

NSwag version: 11.13.3.0
NJsonSchema version: 9.10.22.0 (Newtonsoft.Json v9.0.0.0)

Duration: 00:00:00.0073422
PS C:\Users\joeri\source\repos\nswag1175\nswag1175> dotnet publish -o ./bin/Publish
Microsoft (R) Build Engine version 15.5.180.51428 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 88.29 ms for C:\Users\joeri\source\repos\nswag1175\nswag1175\nswag1175.csproj.
  Restore completed in 76.86 ms for C:\Users\joeri\source\repos\nswag1175\nswag1175\nswag1175.csproj.
  nswag1175 -> C:\Users\joeri\source\repos\nswag1175\nswag1175\bin\Debug\netcoreapp2.0\nswag1175.dll
  nswag1175 -> C:\Users\joeri\source\repos\nswag1175\nswag1175\bin\Publish\
PS C:\Users\joeri\source\repos\nswag1175\nswag1175> nswag run /runtime:NetCore20
NSwag command line tool for .NET 4.6.1+ WinX64, toolchain v11.13.3.0 (NJsonSchema v9.10.22.0 (Newtonsoft.Json v9.0.0.0))
Visit http://NSwag.org for more information.
NSwag bin directory: C:\Program Files (x86)\Rico Suter\NSwagStudio\Win

Executing file 'C:\Users\joeri\source\repos\nswag1175\nswag1175\config.nswag'...
System.TypeLoadException: Method 'StartAsync' in type 'nswag1175.MyService' from assembly 'nswag1175, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.

Server stack trace:
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at NSwag.Commands.SwaggerGeneration.WebApiToSwaggerCommand.<GetControllerTypesAsync>d__91.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\SwaggerGeneration\WebApiToSwaggerCommand.cs:line 285
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.SwaggerGeneration.WebApiToSwaggerCommand.<RunIsolatedAsync>d__88.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\SwaggerGeneration\WebApiToSwaggerCommand.cs:line 188
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.IsolatedCommandBase`1.IsolatedCommandAssemblyLoader`1.Run(String commandType, String commandData, String[] assemblyPaths, String[] referencePaths) in C:\projects\nswag\src\NSwag.Commands\Commands\IsolatedCommandBase.cs:line 59
   at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NSwag.Commands.IsolatedCommandBase`1.IsolatedCommandAssemblyLoader`1.Run(String commandType, String commandData, String[] assemblyPaths, String[] referencePaths)
   at NSwag.Commands.IsolatedCommandBase`1.<>c__DisplayClass13_0.<RunIsolatedAsync>b__0() in C:\projects\nswag\src\NSwag.Commands\Commands\IsolatedCommandBase.cs:line 49
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.IsolatedCommandBase`1.<RunIsolatedAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.IsolatedSwaggerOutputCommandBase.<RunAsync>d__4.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\IsolatedSwaggerOutputCommandBase.cs:line 24
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.NSwagDocumentBase.<GenerateSwaggerDocumentAsync>d__40.MoveNext() in C:\projects\nswag\src\NSwag.Commands\NSwagDocumentBase.cs:line 266
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.NSwagDocument.<ExecuteAsync>d__8.MoveNext() in C:\projects\nswag\src\NSwag.Commands\NSwagDocument.cs:line 76
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.Document.ExecuteDocumentCommand.<ExecuteDocumentAsync>d__9.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\Document\ExecuteDocumentCommand.cs:line 77
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NSwag.Commands.Document.ExecuteDocumentCommand.<RunAsync>d__8.MoveNext() in C:\projects\nswag\src\NSwag.Commands\Commands\Document\ExecuteDocumentCommand.cs:line 43
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.<ProcessSingleAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.<ProcessAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NConsole.CommandLineProcessor.Process(String[] args, Object input)
   at NSwag.Commands.NSwagCommandProcessor.Process(String[] args) in C:\projects\nswag\src\NSwag.Commands\NSwagCommandProcessor.cs:line 54
PS C:\Users\joeri\source\repos\nswag1175\nswag1175>
```

**Comment by RSuter** [comment]
The problem here is that the `/runtime` only works with the NPM package of NSwag - but in your case the NSwagStudio MSI installer installation is used where you cannot change the runtime via cmd line (only the UI). Running this should work for you (just a temporary fix): 

    dotnet "C:\Program Files (x86)\Rico Suter\NSwagStudio\NetCore20\dotnet-nswag.dll" run

**Comment by RSuter** [comment]
Btw: This line shows that you are still running the full .NET version even if you used /runtime: 

![image](https://user-images.githubusercontent.com/2603405/35853541-13dfcf60-0b2e-11e8-91cc-80b6e11b32b7.png)

**Comment by HB-2012** [comment]
If it is a temporary fix, I'd rather use the npm package. I'm not specifically tied to the UI, it just looked like the easiest way to get the CLI at the time (little did I know..). I just tried the npm package, and it indeed works. Thanks!

By the way, would there be a possibility to save the runtime in the config.nswag? I generated the configuration file using NSwagStudio and I didn't know I had to specify the runtime on the command line separately. Especially because the runtime option is part of the same UI as all the other settings.

**Comment by HB-2012** [comment]
Also, some kind of error message for an unsupported runtime would probably be helpful in this scenario.

**Comment by RSuter** [comment]
If you set the runtime in UI then the command line checks whether the runtime is correct and prints an error otherwise... If you leave "Default" then it will run with the current runtime...

**Comment by HB-2012** [comment]
I'll remember that for next time. That however doesn't take away that I didn't receive any notification about the unsupported/missing runtime when I used the CLI that came with the MSI. The documentation also doesn't state it. Through the Wiki I ended up on [this](https://github.com/RSuter/NSwag/wiki/CommandLine) wiki page, from which I then proceeded to install the NSwagStudio using Chocolatey. I previously said I generated the config file using NSwagStudio, but I actually didn't even do that. I copied it somewhere from the wiki. So basically I should've just installed the NPM package. I never had the intention to use NSwagStudio since I'm integrating it in a build script.

Anyway, thanks for helping out.

**Comment by RSuter** [comment]
The biggest problem with the NSwagStudio MSI installer is that it registers the command line path so that you can use "nwag" in your command line but you cannot use /runtime and this is not expected... Maybe we should not register the command line path and use the NSwagStudio MSI installer just for the UI. And if you need the command line tools you can either download them manually or use NPM alongside NSwagStudio. Would this make it easier to use the tools?

**Comment by HB-2012** [comment]
TL;DR: Probably, but updating the documentation would already make a huge difference.

I think the CLI and NSwagStudio are two different products, and from what I feel right now the methods to download these products are kind of mixed up. I use [CakeBuild](https://cakebuild.net) for my build scripts and I found that there was no NuGet package. Since we already use Chocolatey for other things, I opted for that next. 

I didn't expect the CLI in NSwagStudio to be different. This is where the documentation can be more clear and that would already help tremendously. For example on [this](https://github.com/RSuter/NSwag/wiki/Assembly-loading#net-core) wiki section, it isn't explicitly listed that the CLI that comes with NSwagStudio is different. 

And perhaps offering the CLI in a separate NuGet/Chocolatey/NPM package would make NSwag easier to download, but that's just nitpicking.

**Comment by RSuter** [comment]
Just to clarify: The CLI is **not** different - there are four CLI version distributed via NSwagStudio or NPM: Full .NET (x64 and x86), Core 1/1.1/2.0... The only difference is that NSwagStudio calls the correct CLI depending on the Runtime dropdown, the NPM package chooses the CLI depending on the /runtime parameter, but the NSwagStudio installer only registers the /Win (full .NET CLI) and thus the full .NET CLI is always used when installing NSwagStudio and using nswag via CLI (/runtime does not work because it is handled by the NPM package)...

**Comment by RSuter** [comment]
FYI: I added a nswag.cmd file to the NSwagStudio installer so that you can use /runtime also if NSwag is installed via MSI installer: https://github.com/RSuter/NSwag/commit/04576e4d1b714ac5bd9b5df11d72469a23f7c0ff#diff-66d5022663832f3a9a2b8cc682a9ef83
